### ISSUE TITLE
TextTheme TextStyle name changes should have come with a Refactoring Tool

### ISSUE THREAD

**Issue by feinstein** [flutter/flutter#58033]
This is more of a Post-Mortem + Feature Request for future updates than an actual bug report.

The change that deprecated most of the names of the TextTheme TextStyles was https://github.com/flutter/flutter/pull/48547.

I don't mind breaking changes in Flutter, I even encourage them. I believe we should make them as soon as possible while the Framework is young so we don't accumulate a huge list of not ideal APIs for "historical reasons".

That being said, I also believe that those breaking changes should try to be as transparent as possible, leveraging refactoring tools as much as they can.

RxDart for [example](https://github.com/ReactiveX/rxdart#upgrading-from-rxdart-022x-to-023x) made a [tool](https://github.com/brianegan/rxdart_codmod) for automating refactoring for users upgrading from version 0.22.x to 0.23.x.

I upgraded to Flutter 1.17.1 with the new `Typography` API and this is how my Dart Analysis looks like:

![image](https://user-images.githubusercontent.com/8120472/82939364-a4978f80-9f69-11ea-9936-bc5d03e5438f.png)

Sure I can leverage some smart `Find and Replace` but still, it would be nicer to have such a tool already shipped along with the breaking changes.

**Comment by HansMuller** [comment]
I agree with you. I've upgraded many internal applications per the deprecation warnings using a few emacs functions and some shell hacks to turn the warnings into lisp data.  For large scale changes like this, it really would be better to provide a tool that anyone could use, that handled the donkey work. Maybe there's a way to get the IDE to do the work.

CC @devoncarew

**Comment by feinstein** [comment]
@HansMuller yes, the IDE plugins could work as well.

Just want to point out that my intention is more to promote this mentality into the Flutter Release lifecycle, so future Breaking Changes could be shipped with a matching tool as much as possible.

**Comment by devoncarew** [comment]
We have plans for a `flutter fix` tool, and associated IDE warnings and quick fixes, and are still in the design stage. This change is definitely a good use case for us.

The fixes would likely be driven by a data-driven declarative file in a package; the package authors could add an entry when making a breaking change, or when deprecating a field or class, and that entry in the file would provide the necessary information to allow tooling to provide warnings to users, and either migration instructions, or, to perform automated code refactorings to update to the newer APIs.

cc @bwilkerson

**Comment by HansMuller** [comment]
FTR: the change that deprecated most of the names of the TextTheme TextStyles was https://github.com/flutter/flutter/pull/48547. I updated this issue's description.

**Comment by feinstein** [comment]
Maybe Dart's `@Deprecated` annotation could be expanded to include more information that could help out in this process? Maybe accepting a new field called `updater` that's a code-gen class that could automate this? Dart could provide the most common updaters, like `RegexUpdater`, `SimpleSubstitutionUpdater`. I know this isn't the right repo to suggest this, I am just floating the idea for now.

**Comment by devoncarew** [comment]
We have at times discussed adding additional metadata to the deprecated annotation; I think some simple declarative syntax could help with the more common 80/20 cases. A declarative file (like in the lib/ folder of a package) would be able to handle a super-set of the things you could do w/ declarative info on an annotation; thats likely the path we'll purse (but both ways could provide value).

**Comment by feinstein** [comment]
It makes sense, I just like to share ideas as you never know how they might evolve in a brain storm with other people with different perspectives.

**Comment by bwilkerson** [comment]
One of the problems with an annotation (such as `deprecated`) is that they are removed when the changed declaration is removed. That makes it harder for us to help users that don't update during the deprecation period. The declarative file will live with the package forever.

**Comment by feinstein** [comment]
Excellent point!

**Comment by Piinks** [comment]
I am going to close this since we introduced Flutter Fix (dart fix cli tool) to automatically migrate deprecated code like this. It was also applied to all of the TextTheme TextStyle deprecations retroactively. :)
